hack/gendiscoverycache/gendiscoverycache.go

```diff
@@ -0,0 +1,126 @@
+package main
+
+// Copyright (c) Microsoft Corporation.
+// Licensed under the Apache License 2.0.
+
+import (
+	"fmt"
+	"io/ioutil"
+	"os"
+	"path/filepath"
+	"strings"
+
+	configv1 "github.com/openshift/api/config/v1"
+	configclient "github.com/openshift/client-go/config/clientset/versioned"
+	"github.com/sirupsen/logrus"
+	"golang.org/x/net/context"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/client-go/discovery"
+	"k8s.io/client-go/discovery/cached/disk"
+
+	"github.com/Azure/ARO-RP/pkg/database"
+	"github.com/Azure/ARO-RP/pkg/env"
+	"github.com/Azure/ARO-RP/pkg/metrics/noop"
+	"github.com/Azure/ARO-RP/pkg/util/encryption"
+	utillog "github.com/Azure/ARO-RP/pkg/util/log"
+	"github.com/Azure/ARO-RP/pkg/util/restconfig"
+	"github.com/Azure/ARO-RP/pkg/util/version"
+)
+
+const discoveryCacheDir = "pkg/frontend/kubeactions/discovery/cache"
+
+func run(ctx context.Context, log *logrus.Entry) error {
+	if len(os.Args) != 2 {
+		return fmt.Errorf("usage: %s resourceid", os.Args[0])
+	}
+
+	_env, err := env.NewEnv(ctx, log)
+	if err != nil {
+		return err
+	}
+
+	cipher, err := encryption.NewXChaCha20Poly1305(ctx, _env, env.EncryptionSecretName)
+	if err != nil {
+		return err
+	}
+
+	db, err := database.NewDatabase(ctx, log.WithField("component", "database"), _env, &noop.Noop{}, cipher, "")
+	if err != nil {
+		return err
+	}
+
+	doc, err := db.OpenShiftClusters.Get(ctx, strings.ToLower(os.Args[1]))
+	if err != nil {
+		return err
+	}
+
+	restconfig, err := restconfig.RestConfig(_env, doc.OpenShiftCluster)
+	if err != nil {
+		return err
+	}
+
+	cvClient, err := configclient.NewForConfig(restconfig)
+	if err != nil {
+		return err
+	}
+
+	cli, err := disk.NewCachedDiscoveryClientForConfig(restconfig, discoveryCacheDir, "", 0)
+	if err != nil {
+		return err
+	}
+
+	err = validateClusterVersion(cvClient)
+	if err != nil {
+		return err
+	}
+
+	return writeAssets(cli, discoveryCacheDir)
+}
+
+func validateClusterVersion(cvClient configclient.Interface) error {
+	cv, err := cvClient.ConfigV1().ClusterVersions().Get("version", metav1.GetOptions{})
+	if err != nil {
+		return err
+	}
+
+	var actualVersion string
+	for _, history := range cv.Status.History {
+		if history.State == configv1.CompletedUpdate {
+			actualVersion = history.Version
+			break
+		}
+	}
+
+	if actualVersion != version.OpenShiftVersion {
+		return fmt.Errorf(
+			"cluster version is %s, but you must generate cache using %s cluster",
+			actualVersion,
+			version.OpenShiftVersion,
+		)
+	}
```

Comment:
I'm ambivalent about this check - I'd be more inclined to join this to writeAssets and write the returned version to assets_version.  Keep the TestVersion unit test, of course.

Reply:
How we gonna deal with delta when same RP talks to 4.3 and 4.4 clusters?

Reply:
@mjudeikis you mean when your 4.4 PR lands and we support multiple versions?

We will have two options:

1. Update this code to cache for each supported verion
2. Replace `version.OpenShiftVersion` references with one of the supported versions (lastet, I think) version and have cacahe only for one version.

I think the second option is reasonable given the fact that the cache is fallback. So the discovery will first try to make a real request. If the real request fails, we will try to read cache. If there is no cache - we error.

This means that if, let's say, in 4.4 some deprecated resoruce got removed and RP is talking to 4.3 cluster - we wil have cache for most of the resources, but not for the one that still exists 4.3, but not in 4.4 (so no cache). We will make a real discovery request to 4.3 API server.

But implementing support for multiple verions is not going to be very difficult. Generating a cache will be a bit of of pain (create a cluster for each versions, generate cache, etc)

Reply:
@jim-minter the reason why I added this check as well as `TestVersion` is go give early feedback that a human is running the command against a wrong version (vs waiting for test/ci to run).

But given that we are not going to run it frequently, I'm happy to get rid of this check